Fix benchmark and add latency breakdown

arun-koshy opened the pull request:
Standardized posix time calculation which should fix negative latencies being returned by benchmark when timezone was mismatched

Also included further breakdown of narwhal latencies ([investigation notes](https://www.notion.so/mystenlabs/End2End-Latency-Investigation-027092e6dbd84aaa9784d502a8cd4d6a)) and updated local benchmark parameters, which is used in CI, to match what we are currently using in production.

```
-----------------------------------------
 SUMMARY:
-----------------------------------------
 + CONFIG:
 Faults: 0 node(s)
 Committee size: 4 node(s)
 Worker(s) per node: 1 worker(s)
 Collocate primary and workers: True
 Input rate: 200 tx/s
 Transaction size: 512 B
 Execution time: 60 s

 Header number of batches threshold: 1 digests
 Header maximum number of batches: 1,000 digests
 Max header delay: 2,000 ms
 GC depth: 50 round(s)
 Sync retry delay: 10,000 ms
 Sync retry nodes: 3 node(s)
 batch size: 500,000 B
 Max batch delay: 200 ms
 Max concurrent requests: 500,000 

 + RESULTS:
 Batch creation latency: 179 ms
 Header creation latency: 462 ms
        Batch to header latency: 231 ms
 Certificate creation latency: 13 ms
        Request vote outbound latency: 6 ms
 Certificate commit latency: 460 ms

 Consensus TPS: 159 tx/s
 Consensus BPS: 81,616 B/s
 Consensus latency: 472 ms

 End-to-end TPS: 159 tx/s
 End-to-end BPS: 81,414 B/s
 End-to-end latency: 604 ms
-----------------------------------------
```

mwtian reviewed (commented):
Sorry for the delay. I'm looking at the PR. My main question is how do you interpret the sub-component latencies that do not sum to the e2e latency. Still looking at the individual breakdowns.

mwtian commented on narwhal/worker/src/batch_maker.rs:
```diff
@@ -267,6 +267,15 @@ impl BatchMaker {
 
+        #[cfg(feature = "benchmark")]
+        // NOTE: This log entry is used to compute performance.
+        tracing::info!(
```
Can we log at debug level if not in benchmark mode? This will help debugging in production too. Or enable logging at debug level in benchmarks?

arun-koshy replied:
Just added as a debug line that can be used in both prod and in the benchmarks

mwtian commented on narwhal/worker/src/batch_maker.rs:
```diff
@@ -267,6 +267,15 @@ impl BatchMaker {
+        tracing::info!(
+            "Batch {:?} took {} seconds to create due to {}",
+            batch.digest(),
+            self.batch_start_timestamp.elapsed().as_secs_f64(),
```
Can we record the elapsed time and reuse it for both logging and monitoring?

mwtian commented on narwhal/primary/src/proposer.rs:
```diff
@@ -241,7 +241,18 @@ impl Proposer {
+            {
+                // NOTE: This log entry is used to compute performance.
+                tracing::info!(
+                    "Batch {:?} from worker {} took {} seconds to be included in a proposed header",
```
1. Let's mention this is batch creation to inclusion in header?
2. I think this should be calculated as `header.created_at - created_at_timestamp`?

arun-koshy replied:
The two use cases for this timestamp
1. How long does it take for a batch, after creation, to be included in a header?
2. How long does it take for a header to be created?

Will calculate the first as the difference from header creation time to batch creation time. It will come out to the same value because we were calculating the `created_at_timestamp.elapsed_time()` from `now()` at the same time we were generating the header which has its `created_at` time initialized to `now()`

The second will be calculated as the difference of the header `created_at` time and the `created_at_timestamp` of the first batch included in the header.

mwtian commented on narwhal/primary/src/core.rs:
```diff
@@ -479,6 +480,16 @@ impl Core {
+        tracing::info!(
+            "Header {:?} took {} seconds to be materialized to a certificate {:?}",
+            certificate.header.digest(),
+            created_at.elapsed().as_secs_f64(),
```
I think this calculation should be `certificate.metadata.created_at - certificate.header.created_at`, because commit time is calculated from `certificate.metadata.created_at`?

mwtian reviewed (commented):
A general comment is that it will be nice to log these messages at debug level in production, which will be useful for analysis. We already log multiple lines per header and certificate. Logging once per batch should be ok too.

mwtian reviewed (approved):
Feel free to merge and iterate on this.

arun-koshy commented on narwhal/benchmark/fabfile.py:
```diff
@@ -20,15 +20,15 @@ def local(ctx, debug=True):
     }
     node_params = {
-        'header_num_of_batches_threshold': 32,
+        'header_num_of_batches_threshold': 1,
```
We really only need to lower it if tps is lower as well, will move the rate back to 50k and this back to 32 to match

arun-koshy commented:
> My main question is how do you interpret the sub-component latencies that do not sum to the e2e latency

The way the original end to end latency is being calculated takes the earliest timestamp of the sent tx_id and received (committed) tx_id across all validators (which is the most efficient latency). However the latency metrics I added are an average of all nodes latencies broken down by section so I believe that is why there is a discrepancy. 

> it will be nice to log these messages at debug level in production

Agreed, added!



Do you mind checking this [debug line](https://github.com/MystenLabs/sui/blob/b0853de29f0a9b7fc5eff1287f620dd5c4e14a1d/narwhal/primary/src/proposer.rs#L313-L316)? Just want to make sure it makes sense. If so I will merge the PR and iterate from here.

mwtian commented on narwhal/benchmark/benchmark/logs.py:
```diff
@@ -158,12 +198,17 @@ def _parse_workers(self, log):
         tmp = findall(r'Batch ([^ ]+) contains sample tx (\d+)', log)
         samples = {int(s): d for d, s in tmp}
 
+        tmp = findall(
```
Does this still work with logging at debug level?

arun-koshy replied:
Yeah the benchmarks run with debug logs on

mwtian commented on narwhal/benchmark/fabfile.py:
```diff
@@ -22,13 +22,13 @@ def local(ctx, debug=True):
         'header_num_of_batches_threshold': 32,
         'max_header_num_of_batches': 1000,
-        'max_header_delay': '200ms',  # ms
+        'max_header_delay': '5000ms',  # ms
```
I think production value is 2000?

mwtian reviewed (commented):
Ok, we can iterate on the best way to break down latencies. I think there is a way to calculate sub-component latencies so that their sum makes sense.

I was thinking of adding a macro for perf logging, that logs at debug level by default, but info in benchmarks. We can iterate on that as well.

The new log line about leader is a bit inaccurate. Basically the header is actually missing the leader from the last round. And this message will be a bit noisy, since leader will not exist majority of the time. Is it really needed?
